File: leetcode/editor/en/[456]132 Pattern.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var find132pattern = function (nums) {
    if (nums.length < 3) {
        return false;
    }

    let second = Number.MIN_SAFE_INTEGER;
    let stack = [];

    for (let i = nums.length - 1; i >= 0; i--) {
        if (nums[i] < second) {
            return true;
        }

        while (stack.length > 0 && stack[stack.length - 1] < nums[i]) {
            second = stack.pop();
        }

        stack.push(nums[i]);
    }

    return false;
};
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[456]132 Pattern.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public boolean find132pattern(int[] nums) {
        if (nums.length < 3) {
            return false;
        }

        Deque<Integer> dq = new ArrayDeque<>();
        int second = Integer.MIN_VALUE;

        for (int i = nums.length - 1; i >= 0; i--) {
            if (nums[i] < second) {
                return true;
            }

            while (!dq.isEmpty() && dq.peek() < nums[i]) {
                second = dq.pop();
            }

            dq.push(nums[i]);
        }

        return false;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
